injector/mapping.go

```diff
@@ -0,0 +1,143 @@
+	mapping = mapping.DeepCopy()
+
+	if mapping.Annotations == "" {
+		mapping.Annotations = ".spec.template.metadata.annotations"
+	}
+	if len(mapping.Containers) == 0 {
+		mapping.Containers = []servicebindingv1alpha3.ClusterWorkloadResourceMappingContainer{
+			{
+				Path: ".spec.template.spec.initContainers[*]",
+				Name: ".name",
+			},
+			{
+				Path: ".spec.template.spec.containers[*]",
+				Name: ".name",
+			},
+		}
+	}
+	for i := range mapping.Containers {
+		c := &mapping.Containers[i]
+		if c.Env == "" {
+			c.Env = ".env"
+		}
+		if c.VolumeMounts == "" {
+			c.VolumeMounts = ".volumeMounts"
+		}
+	}
+	if mapping.Volumes == "" {
+		mapping.Volumes = ".spec.template.spec.volumes"
+	}
+
+	return mapping
+}
+
+var _ MappingSource = (*mappingSource)(nil)
+
+type mappingSource struct {
+	client client.Client
+}
+
+// NewMappingSource creates a MappingSource confired to lookup a ClusterWorkloadResourceMapping against a cluster. The
+// client should typically be backed by an informer for the ClusterWorkloadResourceMapping kind.
+func NewMappingSource(client client.Client) MappingSource {
+	return &mappingSource{
+		client: client,
+	}
+}
+
+func (m *mappingSource) Lookup(ctx context.Context, workload client.Object) (*servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate, error) {
+	gvk, err := apiutil.GVKForObject(workload, m.client.Scheme())
+	if err != nil {
+		return nil, err
+	}
+	rm, err := m.client.RESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, err
+	}
+	wrm := &servicebindingv1alpha3.ClusterWorkloadResourceMapping{}
+	err = m.client.Get(ctx, types.NamespacedName{Name: fmt.Sprintf("%s.%s", rm.Resource.Resource, rm.Resource.Group)}, wrm)
+	if err != nil {
+		if !apierrs.IsNotFound(err) {
+			return nil, err
+		}
+		// use wildcard version default
+		wrm.Spec.Versions = []servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate{
+			{Version: "*"},
+		}
+	}
+
+	// find version mapping
+	var wildcardMapping *servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate
+	var mapping *servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate
+	for _, v := range wrm.Spec.Versions {
+		switch v.Version {
+		case gvk.Version:
+			mapping = &v
+		case "*":
+			wildcardMapping = &v
+		}
+	}
+	if mapping == nil {
+		mapping = wildcardMapping
+	}
+	if mapping == nil {
+		return nil, fmt.Errorf("no matching version found for %q", gvk)
+	}
+
+	return DefaultWorkloadMapping(mapping), nil
+}
+
+var _ MappingSource = (*staticMapping)(nil)
+
+type staticMapping struct {
+	mapping *servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate
+	err     error
+}
+
+// NewStubMapping returns a single ClusterWorkloadResourceMappingTemplate or error for each lookup. It is useful for
+// testing.
+func NewStubMapping(mapping *servicebindingv1alpha3.ClusterWorkloadResourceMappingTemplate, err error) MappingSource {
```

Comment:
Would you conside `gomock` for mocking instead home-grown solution?

Reply:
stubs > mocks

Mocks tend to be very brittle and hard to work with. I generally avoid them unless the behavior is otherwise difficult to test.

That said, I'm not crazy about calling this a stub vs a static mapping, I'll rename it

Reply:
> Mocks tend to be very brittle and hard to work with.

My experience with `gomock` library is simply great and I got rid of ton of custom code that I need to manage, especially when API evolves.